A81-xds-authority-rewriting.md

```diff
@@ -0,0 +1,171 @@
+fairly general, since it may be used to trigger other security-sensitive
+functionality in the future.)
+
+### xDS Resource Validation
+
+When validating an RDS resource, if the `trusted_xds_server`
+server option is present for this xDS server in the
+bootstrap config, we will look at the [`RouteAction.auto_host_rewrite`
+field](https://github.com/envoyproxy/envoy/blob/b65de1f56850326e1c6b74aa72cb1c9777441065/api/envoy/config/route/v3/route_components.proto#L1173)
+field.  The boolean value of this field will be included in the parsed
+resource struct that is passed to the XdsClient watcher.
+
+When validating an EDS resource, we will look at the [`Endpoint.hostname`
+field](https://github.com/envoyproxy/envoy/blob/b65de1f56850326e1c6b74aa72cb1c9777441065/api/envoy/config/endpoint/v3/endpoint_components.proto#L89).
+The string value of this field will be included in a per-endpoint resolver
+`hostname` attribute in the parsed resource struct that is passed to the
+XdsClient watcher.  Note that the resolver attribute used here should
+be a general-purpose one, not something specific to EDS; for example, in
+the future, we may want to use this same attribute to expose per-endpoint
+hostnames in channelz (see [gRFC A14][A14]).
+
+For Logical DNS clusters (see [gRFC A37][A37]), the same `hostname`
+resolver attribute will be added to all endpoints.  It will be set to
+the name that is resolved for the Logical DNS cluster.
+
+### xDS ConfigSelector Changes
+
+When the xDS ConfigSelector performs routing, it will need to pass
+the values of the route's `auto_host_rewrite` field to the LB picker.
+This data will be passed using the same mechanism introduced in [A31] to
+pass the cluster name to the xds_cluster_manager LB policy.  Note that
+if the implementation is already passing along all of the information
+about the selected route as per [A60], then no changes may be needed here.
+
+### xds_cluster_impl LB Policy Changes
+
+The xds_cluster_impl policy will store the value of the `hostname`
+attribute in the subchannel wrapper when its child policy creates a
+subchannel.  In the xds_cluster_impl picker, if the `auto_host_rewrite`
+option is enabled in the route (passed from the ConfigSelector as
+described above) and the `hostname` attribute on the subchannel wrapper
+is non-empty, the picker will set the `:authority` header to the value of
+the `hostname` attribute.
+
+To support this, we will add an API to allow the LB picker to explicitly
+set the `:authority` header for the RPC.  Note that this authority
+rewriting will override the default authority for the subchannel, which
+would otherwise be set from the resolver factory, a channel option,
+or a per-address attribute returned by the resolver (i.e., the xDS
+authority rewriting will take precedence over any of those settings).
+However, if the implementation allows the application to explicitly
+set the authority on a per-RPC basis (currently only C-core does this),
```

Comment:
@kannanjgithub, note that Java _partly_ has this (`CallOptions.withAuthority`). But very importantly it is missing the next paragraph. Whether we need to just start enforcing the hostname checking in the current API or whether we should make a new API and deprecate+remove the old API is TBD.